KAFKA-3910 [WIP]: Cyclic schema support in ConnectSchema and SchemaBuilder

johnhofman opened the pull request:
This feature uses a FutureSchema as a placeholder to be resolved later. Resolution is attempted whenever a ConnectSchema is constructed, it attempts to resolve all its children (fields, keySchema, or valueSchema) and recurses until the end of the tree. 

A FutureSchema is resolved when it finds a parent schema that matches its name, and optional flag. If a FutureSchema is accessed before being resolved, it will throw a DataException.

The SchemaBuilder constructs a FutureSchema if a field is added with only a type name.

johnhofman committed:
Add FutureSchema to support cyclic schemas

johnhofman committed:
Fix resolution failure test

johnhofman committed:
Move cyclic comparison resolution to FutureSchema

johnhofman committed:
Clean up unused tokens, minor spelling fixes

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/FutureSchema.java:
```diff
@@ -0,0 +1,151 @@
+import java.util.Map;
+import java.util.Objects;
+
+public class FutureSchema implements Schema {
```
Does this actually need to be public? It looks like the SchemaBuilder uses names anyway and expects resolution to work internally. Adding another Schema class seems like its going to get confusing.

ewencp commented on connect/api/src/main/java/org/apache/kafka/connect/data/SchemaBuilder.java:
```diff
@@ -325,6 +325,29 @@ public SchemaBuilder field(String fieldName, Schema fieldSchema) {
+     * @param optional a flag if the schema is optional or not
+     * @return the SchemaBuilder
+     */
+    public SchemaBuilder field(String fieldName, String schemaName, boolean optional) {
```
Wouldn't you also need to handle default values? Presumably you're also assuming that there's only one version of a schema referenced anywhere in the schema? And that the schema parameters (e.g. the scale used by Decimal) don't matter? I realize that Decimal isn't a good example of where this could come into play, I'm just trying to highlight that I'm not sure the current API really addresses the full problem.

ewencp commented:
@johnhofman Only gave it a quick once over, but can I suggest a possible alternative implementation? `SchemaBuilder` intentionally implements the `Schema` interface. Could we just make the `build()` method handle resolving any `SchemaBuilders` into their final `Schemas`? We pass in unmodifiable maps, but that isn't strictly required (it's just to add some level of protection to the schemas), so I think you could generate the output structure and then patch things up if there are any references to `SchemaBuilders` in there.

johnhofman commented:
@ewencp Yeah, I started in that direction, resolving the references just using `ConnectSchema` before finishing the build, but ran into a few issues. The `ConnectSchema` also sets it's key and value schemas  as final in the constructor, so patching them up afterwards when it's an array or map would need us to be flexible on that `final` as well. But the blocking issue was that once the schema is built,  doing `.equals` or `.hashcode` causes a loop. The `FutureSchema` knows to break the hashing/equals loop. I have since thought of some scenarios where the current logic will not work as desired (for example starting a comparison with the `FutureSchema` as root), but have fix in progress (WIP until then).

In regards to the access of `FutureSchema`, I think you are correct that it shouldn't be visible outside the `SchemaBuilder`.  I now understand that the optional and default fields are related to the 'instance' of a type rather than the definition of the type. Which category are `parameters` in?

I think extending the `SchemaBuilder` with a `private FutureSchemaBuilder` that only allows setting `optional` and `default` (maybe `parameters`?) clean up the API. Thoughts?

ewencp commented:
@johnhofman Not sure I understand the issue w/ `final` re: map key and value schemas. How does this differ at all from self-referential schemas where the contents of the schema need to be mutable but the reference to the schema object itself doesn't need to be? And `equals` and `hashcode` issues seem like they can be resolved by checking for identity equality to break these loops? Or is the issue that we need to track multiple schemas if schemas aren't directly self-referential but instead are indirectly self-referential?

I'm not quite sure what you mean by the type vs instance of the type, but I think in your terms `parameters` are really about the instance of the type. The simplest example is the `Decimal` logical type -- it has a `scale` parameter that controls the decimal precision. It's tied to the instance of the type since we can have many different schemas of type `Decimal`, but they can differ in precision.

johnhofman committed:
Make FutureSchema private and use context for comparisons.

FutureSchema's are built with a FutureSchemaBuilder which is returned when building a type by name. Equals and hash operations on cyclic schema's are resolved by passing a context (list of visited nodes) along during the operation.

johnhofman commented:
@ewencp My concept of `instance` vs `type` (I expect there is correct terminology for this, but I haven't yet come across it) is easily illustrated by comparing an avro schema with it's connect schema equivalent :

```
{"type":"record",
 "name":"avro",
 "fields":[{"name":"int","type":"int"}, 
           {"name":"int_optional","type":["null","int"]},
           {"name":"int_with_default","type":"int","default":"0"}]
}
```

The record `avro` has three fields, all different `instances` of the type int. The three fields all reference the same type, however one instance is optional, one instance has a default value and the other is just a plain int.

As a `ConnectSchema` this would be constructed:

```
SchemaBuilder.struct().name("connect")
  .field("int", Schema.INT32_SCHEMA)
  .field("int_optional", Schema.OPTIONAL_INT32_SCHEMA)
  .field("int_with_default", SchemaBuilder.int32().defaultValue(0).build())
.build();
```

This represents a record with the same fields, except here the "optionality" and "default" are embedded in the schema so that each field actually references a different schema object. This means that we need the `FutureSchema` as a wrapper that can reference an existing schema (to reuse the type defining properties) but can override the `optional` and `default` properties. 

Re: equality and hashing: The tricky part is indeed indirectly cyclic schema's, and making the solution thread-safe. The only solution I can think of is to track the visited nodes and break the cycle when revisiting a node. To do this thread-safe a "context" needs to be passed along with the comparison, the resulting code is in the latest change set.
